Feat: 화장실 검색 (#7)

hyuntae99 opened the pull request:
### 🔖작업 내용

1. 회원 아이디, 검색어, 검색 결과 등을 저장하는 search entity 생성
2. t-map api를 이용하여 검색어에 해당하는 도로명 주소, 지번 주소를 추출 후 저장
3. 회원 아이디와 검색어를 받는 검색 기능과 클릭 처리, 최근 검색어 불어오기 구현
4. 예외 처리를 위한 ErrorCode와 ExceptionHandler 생성
(개발하면서 발생한 오류나 많이 쓰는 오류에 대해서 간단히 구현)

### 🔖논의 사항
1. id (Integer -> Long) 으로 변경하는게 어떤가?
2. auth_type이 숫자이면 오류가 안나는데 문자이면 NumberFormatException 오류가 발생함 (확인 필요)
3. 현재 user_id를 요청 받아서 저장하고 있는데 로그인 중인 회원이 검색을 사용하면 요청 받지 않아도 알아서 user_id를 받아올 수 있나?

### 📨관련 이슈

Fixes: https://github.com/chamjimayo/backend/issues/7

hyuntae99 committed:
T-Map API를 이용한 주소 검색 기능 구현
1. 검색어(keyWord)를 통해 도로명 주소를 반환
2. 유저Id와 검색어, 도로명 주소를 search entity에 저장
3. 유저가 선택(클릭)한 주소를 판단하기 위해 click 변수를 도입
4. 클릭 처리한 도로명 주소 중 가장 최근의 주소를 반환
5. application.yml 에 t-map appKey 설정
(보안을 위해서 .gitignore 에 application.yml 추가)

hyuntae99 committed:
1. search Entity에 지번 주소 추가
2. 응답에 지번 주소 추가
3. Dto 이름 변경

hyuntae99 committed:
통합 예외 관리를 위한 ExceptionHandler 추가
(일단 간단한 오류에 대해서만 처리함)

hyuntae99 committed:
1. setter -> builder 사용
2. 통상적으로 많이 발생하는 오류 추가
3. HTTP Method 변경

hyuntae99 committed:
upstream merge

hyuntae99 committed:
upstream merge

hyuntae99 commented:
cherry-pick 이라는 기능을 이용해서 해결했는데 혹시 commit 꼬이면 [참고자료](https://brownbears.tistory.com/606) 보고 해결하세요~

hyuntae99 committed:
화장실 정보 추가

seokho-1116 commented:
1. 굿 Long이 좋다그러더라
2. 어디부분에서?
3. spring security의 filter를 사용해서 user의 jwt 토큰을 받아서 저장해 controller에서 token을 보낸 user의 Id를 쓸 수 있게 할거임

seokho-1116 commented on src/main/java/com/project/chamjimayo/domain/entity/Search.java:
```diff
@@ -38,4 +39,33 @@ public class Search {
+	}
+
+	@Builder
+	public void setClick(int click) {
```
여기는 Builder로 처리하기 보다는 setter라도 의미있는 이름으로 정하는게 좋다고 하던데 changeClick 이런식으로?

seokho-1116 commented on src/main/java/com/project/chamjimayo/exception/ErrorCode.java:
```diff
@@ -0,0 +1,42 @@
+	DATABASE_ERROR(500, "COMMON-005", "데이터 베이스 오류 발생"),
+	INTER_SERVER_ERROR(500, "COMMON-006", "서버 오류 발생");
+
+	private int status;
```
status는 어차피 ResponseEntity에서 정해주기 때문에 안넣어도됨

seokho-1116 commented on src/main/java/com/project/chamjimayo/exception/CustomException.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Getter;
+
+@Getter
+public class CustomException extends RuntimeException {
```
CustomException에 ErrorCode로 분리하기 보단 각각 에러별로 클래스 생성해주는게 이해가 쉬울듯한데 이건 정답은 없어보임

seokho-1116 commented on src/main/java/com/project/chamjimayo/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,20 @@
+@Getter
+public class ErrorResponse {
+
+	private final int status;
```
여기도 status 없어도 됨

seokho-1116 commented on src/main/java/com/project/chamjimayo/security/dto/SearchRequestDto.java:
```diff
@@ -0,0 +1,24 @@
+	private final String searchWord;
+	private final Long userId;
+
+	@Builder
```
Builder 좋지만 매개변수가 적을 때는 정적 메소드 팩토리 패턴으로 받아도 좋아!!
[정적 팩토리 메소드](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

seokho-1116 commented on src/main/java/com/project/chamjimayo/security/dto/SearchResponseDto.java:
```diff
@@ -0,0 +1,21 @@
+	private final String lotNumberAddress;
+	private final String name;
+
+	@Builder
```
Builder 좋지만 매개변수가 적을 때는 정적 메소드 팩토리 패턴으로 받아도 좋아!!
[정적 팩토리 메소드](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)

seokho-1116 commented on src/main/java/com/project/chamjimayo/controller/SearchController.java:
```diff
@@ -0,0 +1,66 @@
+	private final SearchService searchService;
+	private final SearchRepository searchRepository;
+
+	@Autowired
```
롬복쓰고 클래스 변수들이 final이니까 required args constructor 애노테이션 어떰?
롬복안쓰면 이렇게 하는게 맞음!
[required args constructor](https://medium.com/webeveloper/requiredargsconstructor-%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-%EC%9D%98%EC%A1%B4%EC%84%B1-%EC%A3%BC%EC%9E%85-dependency-injection-4f1b0ac33561)

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/SearchService.java:
```diff
@@ -0,0 +1,211 @@
+	@Value("${tmap.appKey}")
+	private String tmapApikey;
+
+	@Autowired
```
롬복쓰고 클래스 변수들이 final이니까 required args constructor 애노테이션 어떰?
롬복안쓰면 이렇게 하는게 맞음!
[required args constructor](https://medium.com/webeveloper/requiredargsconstructor-%EB%A5%BC-%EC%9D%B4%EC%9A%A9%ED%95%9C-%EC%9D%98%EC%A1%B4%EC%84%B1-%EC%A3%BC%EC%9E%85-dependency-injection-4f1b0ac33561)

seokho-1116 commented on src/main/java/com/project/chamjimayo/exception/ErrorCode.java:
```diff
@@ -0,0 +1,42 @@
+public enum ErrorCode {
+
+	//400 BAD_REQUEST 잘못된 요청
+	INVALID_INPUT_VALUE(400, "COMMON-001", "유효성 검증에 실패한 경우"),
```
오류 코드를 따로 정의해봐야 할 듯?

seokho-1116 commented on .gitmodules:
```diff
@@ -1,3 +1,6 @@
 [submodule "src/main/resources/config"]
 	path = src/main/resources/config
 	url = https://github.com/chamjimayo/config
+[submodule "config"]
```
이거 path 왜바꾼거?
src/main/resources/config에 있어야 resources/application.yml이 config/application-local.yml을 가르켜 
submodule을 쓸 수 있음

seokho-1116 commented on src/main/java/com/project/chamjimayo/service/SearchService.java:
```diff
@@ -0,0 +1,211 @@
+	 */
+	public ResponseEntity<String> clickAddress(Long searchId) {
+		// searchId로 search 를 받아옴
+		Optional<Search> searchOptional = searchRepository.findById(searchId);
```
여기도 위에 orElseThrow처럼 처리하는게 어떤지? If else 빼고

seokho-1116 reviewed (changes requested):
리뷰 했어!! 고생많았으

hyuntae99 committed:
1. ErrorCode, ErrorResponse 통합
2. 각각 Error 별 클래스 생성
3. Builder -> 정적 팩토리 메소드 사용
4. @RequiredArgsConstructor 사용

hyuntae99 commented:
1. 모든 Entity의 id Integer -> Long
  ERD CLOUD도 반영 (INT -> BIGINT)
2. api 테스트할 때, auth_type에 숫자를 넣으면 오류가 없는데 KAKAO, NAVER를 넣으면 오류 발생
  java.lang.NumberFormatException: For input string: "NAVER"
3. 무슨 말인지 잘 모르겠어서 구현은 못했는데 더 알아볼게

hyuntae99 commented:
그리고 ErrorCode, ErrorResponse에 @Getter 추가했어
추가를 안하면 org.springframework.web.HttpMediaTypeNotAcceptableException: Could not find acceptable representation가 생기더라고 구글링해서 해결은 했는데 맞게 했는지는 모르겠네
[참고자료](https://velog.io/@kai6666/Spring-%EC%8A%A4%ED%94%84%EB%A7%81-%EC%8A%AC%EB%9D%BC%EC%9D%B4%EC%8A%A4-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%A4%91-406-org.springframework.web.HttpMediaTypeNotAcceptableException-Could-not-find-acceptable-representation-%EC%97%90%EB%9F%AC)

hyuntae99 commented:
예외 처리는 일단 러프하게 만든거라서 추가할 부분은 계속 추가해야 할 것 같아
클래스를 만들어서 사용하는 것도 처음이라서 부족한 부분 있으면 말해줘!

hyuntae99 committed:
Error 메세지 수정

seokho-1116 commented:
2. @Enumerated(EnumType.String)을 auth_type에 추가했어야 했는데 까먹은듯... 쏘리

3. AuthenticationPrincipal로 내가 넘겨준 유저 정보 받을 수 있어
[시큐리티 현재 사용자](https://itstory.tk/entry/Spring-Security-%ED%98%84%EC%9E%AC-%EB%A1%9C%EA%B7%B8%EC%9D%B8%ED%95%9C-%EC%82%AC%EC%9A%A9%EC%9E%90-%EC%A0%95%EB%B3%B4-%EA%B0%80%EC%A0%B8%EC%98%A4%EA%B8%B0)

seokho-1116 commented:
> 1. 모든 Entity의 id Integer -> Long
>   ERD CLOUD도 반영 (INT -> BIGINT)
> 2. api 테스트할 때, auth_type에 숫자를 넣으면 오류가 없는데 KAKAO, NAVER를 넣으면 오류 발생
>   java.lang.NumberFormatException: For input string: "NAVER"
> ![image](https://github.com/chamjimayo/backend/assets/101180610/7e35f9bf-c479-4b23-98fe-9314205b144d)
> 
> 3. 아직 구현은 못했는데 더 알아볼게

2. @Enumerated(EnumType.String)을 auth_type에 추가했어야 했는데 까먹은듯... 쏘리

3. AuthenticationPrincipal로 내가 넘겨준 유저 정보 받을 수 있어
[시큐리티 현재 사용자](https://itstory.tk/entry/Spring-Security-%ED%98%84%EC%9E%AC-%EB%A1%9C%EA%B7%B8%EC%9D%B8%ED%95%9C-%EC%82%AC%EC%9A%A9%EC%9E%90-%EC%A0%95%EB%B3%B4-%EA%B0%80%EC%A0%B8%EC%98%A4%EA%B8%B0)

seokho-1116 commented:
> 그리고 ErrorCode, ErrorResponse에 @Getter 추가했어
> 추가를 안하면 org.springframework.web.HttpMediaTypeNotAcceptableException: Could not find acceptable representation가
> 발생하더라고 [구글링](https://velog.io/@kai6666/Spring-%EC%8A%A4%ED%94%84%EB%A7%81-%EC%8A%AC%EB%9D%BC%EC%9D%B4%EC%8A%A4-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%A4%91-406-org.springframework.web.HttpMediaTypeNotAcceptableException-Could-not-find-acceptable-representation-%EC%97%90%EB%9F%AC)해서 해결은 했는데 맞게 했는지는 모르겠네
> 
> 

굿굿 👍

hyuntae99 committed:
User AuthType에 @Enumerated(EnumType.STRING)적용

hyuntae99 committed:
error log 추가

hyuntae99 committed:
가독성을 위한 예외처리 순서 변경

hyuntae99 committed:
UserDetails id (String -> Long)으로 변경

hyuntae99 committed:
@AuthenticationPrincipal 도입
(현재 로그인한 사용자의 userId를 알아서 가져오도록)

hyuntae99 commented:
### 변경 사항
1. User의 auth_type에 @Enumerated(EnumType.String) 적용 
2. @AuthenticationPrincipal 도입 (현재 로그인 된 사용자의 id를 자동으로 받아옴)
3. UserDetails의 id를 String -> Long으로 변경

### 논의 사항
1. UserDetails의 id가 UserId가 맞는가? (맞다고 판단해서 일단 String에서 Long으로 수정)
2.  @AuthenticationPrincipal 도입 후 API 테스트를 해보고 싶은데 
"유효하지 않은 토큰입니다." 라는 오류가 발생해서 맞게 구현한건지 모르겠어 (해결방법 아시는 분)

hyuntae99 committed:
프로젝트 구조 변경

seokho-1116 commented:
1. security/CustomUserDetails.java라면 id가 userId맞음
2. 이건 Authorization 헤더에 올바른 jwt를 넣어야 해

hyuntae99 committed:
BaseEntity 활성화를 위한 Annotation 수정

hyuntae99 committed:
Swagger 사용을 위한 코드 추가

hyuntae99 committed:
Swagger 사용을 위한 코드 추가

seokho-1116 reviewed (approved):
승인완료 수고했어

hyuntae99 committed:
Swagger 의존성 변경

hyuntae99 committed:
Swagger 코드 설명 추가

hyuntae99 committed:
Swagger 코드 설명 추가

hyuntae99 committed:
1. Swagger ErrorResponse 추가
2. ErrorCode 추가 및 수정

hyuntae99 commented:
수고했어!
내가  swagger 처음 써봐서 그런데 [이렇게](https://github.com/chamjimayo/backend/pull/16/commits/c215e1b134ea6ac616605ca6f1ab200f3a2aa2b3) 구현하는거 괜찮아??
swagger 문서에는 잘나오는 것 같긴 한데

hyuntae99 commented:
일단 기본적인 swagger 코드는 모두 작성했어

seokho-1116 commented:
> 일단 기본적인 swagger 코드는 모두 작성했어

SwggerConfig에는 공통으로 발생할 수 있는 에러만 명시하고 각 컨트롤러 엔드포인트에 대한건 @ApiResponse로 받아야함
찾아보면 나옴

hyuntae99 committed:
swagger 충돌로 인한 HttpStatus 변경

hyuntae99 committed:
1. SwaggerConfig에 공통 error만 반영
2. 각 엔드포인트에 대한 @ApiResponse 추가
3. Dto에 대한 설명 추가

hyuntae99 commented:
1. SwaggerConfig에 공통 Error만 반영 (아닌 것 같다면 바로 말해줘)
2. 각 엔드 포인트에 대한  @ApiResponse 추가 
(Swagger가 같은 HttpStatus에 대해서 처리를 못하더라고 그래서 or로 일단 표기하긴 했는데 혹시 좋은 방법 알면 알려줘)
3. Example Value에 보여줄 Dto에 대한 설명 추가

hyuntae99 committed:
Dto에 대한 설명 추가

hyuntae99 committed:
가독성을 위한 @ApiResponse 수정

hyuntae99 commented:
말한 방법 적용하니까 가독성이 더 좋아진듯
![image](https://github.com/chamjimayo/backend/assets/101180610/6fec1cbf-afed-4763-b472-919fd74ac7f1)

hyuntae99 committed:
upstream merge

hyuntae99 commented:
스키마(hyuntae) 만들어서 테스트 해보려 하는데 회원 가입이 안되네... 
내가 뭐 놓치고 있는게 있나??

hyuntae99 committed:
주석 수정

seokho-1116 commented:
> 스키마(hyuntae) 만들어서 테스트 해보려 하는데 회원 가입이 안되네... 내가 뭐 놓치고 있는게 있나??

회원가입 안됨? 어떤 오류 떠?

hyuntae99 committed:
엔드포인트 추가

hyuntae99 commented:
1. 회원가입 / 로그인은 해결했어!
2. @AuthenticationPrincipal를 사용하기 위해서는 CustomUserDetails에 대한 정보가 필요한데
어떤 데이터를 넣어줘야 할지 모르겠네...
![image](https://github.com/chamjimayo/backend/assets/101180610/6e16fdf8-38b7-4c30-87ce-a6f45537edab)

kick-sim commented on src/main/java/com/project/chamjimayo/security/CustomUserDetails.java:
수고하셨슴다!!
일부러 id 밖에 사용 안해서 password랑 name은 return null 로 해놓은거지?

hyuntae99 replied:
나도 CustomUserDetails에 어떤 데이터가 들어가는지 모르겠어

hyuntae99 committed:
swagger response 수정 및 추가

hyuntae99 committed:
error 코드 수정

hyuntae99 committed:
코드 수정
